pubtools/exodus/_hooks/pulp.py

```diff
@@ -0,0 +1,210 @@
+        self._session = None
+        self._publish = None
+
+    def _exodus_enabled(self):
+        enable_vals = ["true", "t", "1", "yes", "y"]
+        return os.getenv("EXODUS_ENABLED", "False").lower() in enable_vals
+
+    def _new_session(self):
+        retry_strategy = Retry(
+            total=self._retries,
+            backoff_factor=1,
+            status_forcelist=[429, 500, 502, 503, 504],
+        )
+        adapter = requests.adapters.HTTPAdapter(max_retries=retry_strategy)
+
+        out = requests.Session()
+        out.cert = (self._gw_crt, self._gw_key)
+        out.mount(self._gw_url, adapter)
+
+        return out
+
+    def _unpack_response(self, response):
+        """Raise if 'response' was not successful.
+
+        This is the same as response.raise_for_status(), merely wrapping it
+        to ensure the body is logged when possible."""
+
+        try:
+            response.raise_for_status()
+        except Exception as outer:
+            try:
+                body = response.json()
+            except:
+                raise outer
+
+            LOG.error("Unsuccessful response from exodus-gw: %s", body)
+            raise
+
+    def _do_request(self, **kwargs):
+        if not self._session:
+            self._session = self._new_session()
+
+        resp = self._session.request(**kwargs)
+        self._unpack_response(resp)
+        return resp
+
+    def _check_cert(self):
+        """Issue request to exodus-gw to identify permissions."""
+
+        if not self._gw_crt:
+            LOG.debug("EXODUS_GW_CERT not found, authentication may fail")
+
+        if not self._gw_key:
+            LOG.debug("EXODUS_GW_KEY not found, authentication may fail")
+
+        auth_url = urljoin(self._gw_url, "/whoami")
+        resp = self._do_request(method="GET", url=auth_url)
+        context = resp.json()
+
+        for user_type, ident in (
+            ("client", "serviceAccountId"),
+            ("user", "internalUsername"),
+        ):
+            typed_ctx = context[user_type]
+            if typed_ctx["authenticated"]:
+                LOG.debug(
+                    "Authenticated with exodus-gw at %s as %s %s (roles: %s)",
+                    self._gw_url,
+                    user_type,
+                    typed_ctx[ident],
+                    [str(role) for role in typed_ctx["roles"]],
+                )
+                break
+        else:
+            LOG.debug("Not authenticated with exodus-gw.")
+
+    def _poll_commit_completion(self, commit, tries=4):
+        """Issues request(s) to exodus-gw to get the task state, returning once
+        it either is FAILED or COMPLETED."""
+        task_url = urljoin(self._gw_url, commit["links"]["self"])
+
+        for i in range(tries):
+            resp = self._do_request(method="GET", url=task_url)
+            task = resp.json()
+
+            if task["state"] not in ("FAILED", "COMPLETE"):
+                if i < tries - 1:
```

Comment:
The timing here seems a bit aggressive.  I haven't run all the numbers, but with tries=4 and the wait_time calculation here, aren't we only giving it a few minutes to complete?

That doesn't seem long enough - yes we do expect individual commit tasks to be quick, but it is also possible for exodus-gw to have other tasks in the queue, or for the worker to be temporarily offline in which case we want to give it some time to come back online.

Could we bump this up a bit so it can wait for longer?  And it would be good for this to be somehow configurable as well (another env var?).

Personally, I would not write this by combining a "tries" and "wait_time" because it becomes a little tricky to figure out exactly how it behaves.  I would write this as something like: set a deadline of now + 15 minutes when the commit task is created, then in the loop just do "while now < deadline".